File: scripts/semantic_seam_status.py
#!/usr/bin/env python3
"""Print a JSON status object for the Axiomurgy ↔ Vermyth semantic recommendation seam (read-only).

Reads the committed baseline and corpus; optionally probes HTTP healthz (--probe) or attaches
multi_match_rate from a prior --calibrate JSON report (--calibration-report).

Environment (optional):
  AXIOMURGY_VERMYTH_BASE_URL or VERMYTH_BASE_URL — for --probe / --live

Examples:
  python scripts/semantic_seam_status.py
  python scripts/semantic_seam_status.py --probe
  python scripts/semantic_seam_status.py --calibration-report docs/reports/last_run.json
"""

from __future__ import annotations

import argparse
import importlib.util
import json
import os
import sys
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parents[1]
if str(ROOT) not in sys.path:
    sys.path.insert(0, str(ROOT))


def _load_eval_module():
    path = ROOT / "scripts" / "eval_semantic_recommendations.py"
    spec = importlib.util.spec_from_file_location("eval_semantic_recommendations", path)
    assert spec and spec.loader
    mod = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(mod)
    return mod


def _base_url() -> str | None:
    v = os.environ.get("AXIOMURGY_VERMYTH_BASE_URL") or os.environ.get("VERMYTH_BASE_URL")
    return v.strip().rstrip("/") if isinstance(v, str) and v.strip() else None


def main() -> int:
    ap = argparse.ArgumentParser(description="Semantic seam status JSON (baseline + corpus + optional probe).")
    ap.add_argument(
        "--baseline",
        type=Path,
        default=ROOT / "docs" / "reports" / "compatibility_baseline_live_v1.json",
        help="Committed compatibility baseline JSON",
    )
    ap.add_argument(
        "--corpus",
        type=Path,
        default=ROOT / "docs" / "data" / "semantic_recommend_corpus.json",
        help="Semantic recommend corpus JSON",
    )
    ap.add_argument(
        "--calibration-report",
        type=Path,
        default=None,
        metavar="PATH",
        help="Optional prior harness JSON (with calibration.metrics) for multi_match_rate",
    )
    ap.add_argument(
        "--probe",
        action="store_true",
        help="If AXIOMURGY_VERMYTH_BASE_URL is set, GET /healthz (cheap; no recommend calls)",
    )
    ap.add_argument(
        "--live",
        action="store_true",
        help="With --probe, same as --probe (reserved for future full probe; currently identical)",
    )
    args = ap.parse_args()
    _ = args.live  # reserved

    out: dict[str, Any] = {"kind": "semantic_seam_status", "axiomurgy_repo_root": str(ROOT)}

    if not args.baseline.is_file():
        out["error"] = f"baseline not found: {args.baseline}"
        print(json.dumps(out, indent=2))
        return 2

    baseline = json.loads(args.baseline.read_text(encoding="utf-8-sig"))
    out["baseline_version"] = baseline.get("baseline_version")
    out["baseline_captured_at"] = baseline.get("captured_at")
    out["baseline_axiomurgy_git"] = baseline.get("axiomurgy_git")
    out["baseline_vermyth_git"] = baseline.get("vermyth_git")
    out["baseline_expectations_count"] = len(baseline.get("expectations") or [])

    if args.corpus.is_file():
        corpus = json.loads(args.corpus.read_text(encoding="utf-8-sig"))
        out["corpus_version"] = corpus.get("version")
        out["corpus_spell_count"] = len(corpus.get("spells") or [])

    if args.calibration_report and args.calibration_report.is_file():
        rep = json.loads(args.calibration_report.read_text(encoding="utf-8-sig"))
        cal = rep.get("calibration") or {}
        metrics = cal.get("metrics") or {}
        mm = metrics.get("multi_match_rate")
        if isinstance(mm, dict):
            out["multi_match_rate"] = mm

    if args.probe or args.live:
        base = _base_url()
        if not base:
            out["healthz"] = None
            out["healthz_note"] = "AXIOMURGY_VERMYTH_BASE_URL not set; skipping probe"
        else:
            mod = _load_eval_module()
            hz = mod.fetch_healthz(base, timeout_s=10.0)
            out["healthz"] = hz
            if isinstance(hz, dict) and isinstance(hz.get("status_code"), int):
                out["healthz_status_code"] = hz["status_code"]

    print(json.dumps(out, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: packages/semantic-seam/src/cli/semantic-seam-status.ts
#!/usr/bin/env node
/**
 * TypeScript counterpart to scripts/semantic_seam_status.py (read-only status JSON).
 */
import { readFileSync } from "node:fs";
import { join } from "node:path";
import { fetchHealthz } from "../healthz.js";
import { findRepoRoot } from "../repoRoot.js";

function envBaseUrl(): string | null {
  const v = process.env.AXIOMURGY_VERMYTH_BASE_URL ?? process.env.VERMYTH_BASE_URL;
  if (typeof v === "string" && v.trim()) return v.trim().replace(/\/+$/, "");
  return null;
}

function parseArgs(argv: string[]): {
  baseline: string;
  corpus: string;
  calibrationReport: string | null;
  probe: boolean;
  live: boolean;
} {
  let baseline = join("docs", "reports", "compatibility_baseline_live_v1.json");
  let corpus = join("docs", "data", "semantic_recommend_corpus.json");
  let calibrationReport: string | null = null;
  let probe = false;
  let live = false;
  for (let i = 2; i < argv.length; i++) {
    const a = argv[i];
    if (a === "--probe") probe = true;
    else if (a === "--live") live = true;
    else if (a === "--baseline" && argv[i + 1]) {
      baseline = argv[++i]!;
    } else if (a === "--corpus" && argv[i + 1]) {
      corpus = argv[++i]!;
    } else if (a === "--calibration-report" && argv[i + 1]) {
      calibrationReport = argv[++i]!;
    }
  }
  return { baseline, corpus, calibrationReport, probe, live };
}

async function main(): Promise<number> {
  const repoRoot = findRepoRoot(process.cwd());
  const args = parseArgs(process.argv);
  const baselinePath = join(repoRoot, args.baseline);

  const out: Record<string, unknown> = {
    kind: "semantic_seam_status",
    axiomurgy_repo_root: repoRoot,
  };

  try {
    readFileSync(baselinePath, "utf8");
  } catch {
    out.error = `baseline not found: ${baselinePath}`;
    console.log(JSON.stringify(out, null, 2));
    return 2;
  }

  const baseline = JSON.parse(readFileSync(baselinePath, "utf8")) as Record<string, unknown>;
  out.baseline_version = baseline.baseline_version;
  out.baseline_captured_at = baseline.captured_at;
  out.baseline_axiomurgy_git = baseline.axiomurgy_git;
  out.baseline_vermyth_git = baseline.vermyth_git;
  out.baseline_expectations_count = Array.isArray(baseline.expectations) ? baseline.expectations.length : 0;

  const corpusPath = join(repoRoot, args.corpus);
  try {
    const corpus = JSON.parse(readFileSync(corpusPath, "utf8")) as { version?: unknown; spells?: unknown[] };
    out.corpus_version = corpus.version;
    out.corpus_spell_count = Array.isArray(corpus.spells) ? corpus.spells.length : 0;
  } catch {
    /* optional */
  }

  if (args.calibrationReport) {
    const p = join(repoRoot, args.calibrationReport);
    try {
      const rep = JSON.parse(readFileSync(p, "utf8")) as { calibration?: { metrics?: { multi_match_rate?: unknown } } };
      const mm = rep.calibration?.metrics?.multi_match_rate;
      if (mm !== undefined && typeof mm === "object" && mm !== null) out.multi_match_rate = mm;
    } catch {
      /* ignore */
    }
  }

  if (args.probe || args.live) {
    const base = envBaseUrl();
    if (!base) {
      out.healthz = null;
      out.healthz_note = "AXIOMURGY_VERMYTH_BASE_URL not set; skipping probe";
    } else {
      const hz = await fetchHealthz(base, { timeoutS: 10 });
      out.healthz = hz;
      if (hz && typeof hz.status_code === "number") out.healthz_status_code = hz.status_code;
    }
  }

  console.log(JSON.stringify(out, null, 2));
  return 0;
}

main().then(
  (code) => process.exit(code),
  (err) => {
    console.error(err);
    process.exit(1);
  },
);
